src/integrations.py

```diff
@@ -0,0 +1,250 @@
+    with Operation(dsn) as op:
+        if not op.select(Group, Group.name == group_name):
+            group = Group(name=group_name, gid_number=DEFAULT_GID)
+            op.add(group)
+
+        if not (user := op.select(User, User.name == user_name)):
+            new_password = hashlib.sha256(token_bytes()).hexdigest()
+            user = User(
+                name=user_name,
+                uid_number=DEFAULT_UID,
+                gid_number=DEFAULT_GID,
+                password_sha256=new_password,
+            )
+            op.add(user)
+        password = user.password_bcrypt or user.password_sha256
+
+        if not op.select(Capability, Capability.user_id == DEFAULT_UID):
+            capability = Capability(user_id=DEFAULT_UID)
+            op.add(capability)
+
+    return BindAccount(user_name, group_name, password)
+
+
+class LdapIntegration:
+    def __init__(self, charm: CharmBase):
+        self._charm = charm
+        self._bind_account: Optional[BindAccount] = None
+
+    def load_bind_account(self, user: str, group: str) -> None:
+        database_config = DatabaseConfig.load(self._charm.database_requirer)
+        self._bind_account = _create_bind_account(database_config.dsn, user, group)
+
+    @property
+    def ldap_url(self) -> str:
+        return f"ldap://{self._charm.config.get('hostname')}:{GLAUTH_LDAP_PORT}"
+
+    @property
+    def base_dn(self) -> str:
+        return self._charm.config.get("base_dn")
+
+    @property
+    def provider_base_data(self) -> LdapProviderBaseData:
+        return LdapProviderBaseData(
+            url=self.ldap_url,
+            base_dn=self.base_dn,
+        )
+
+    @property
+    def provider_data(self) -> Optional[LdapProviderData]:
+        if not self._bind_account:
+            return None
+
+        return LdapProviderData(
+            url=self.ldap_url,
+            base_dn=self.base_dn,
+            bind_dn=f"cn={self._bind_account.cn},ou={self._bind_account.ou},{self.base_dn}",
+            bind_password_secret=self._bind_account.password or "",
+            auth_method="simple",
+            starttls=True,
+        )
+
+
+class AuxiliaryIntegration:
+    def __init__(self, charm: CharmBase):
+        self._charm = charm
+
+    @property
+    def auxiliary_data(self) -> AuxiliaryData:
+        database_config = DatabaseConfig.load(self._charm.database_requirer)
+
+        return AuxiliaryData(
+            database=database_config.database,
+            endpoint=database_config.endpoint,
+            username=database_config.username,
+            password=database_config.password,
+        )
+
+
+@dataclass
+class CertificateData:
+    ca_cert: Optional[str] = None
+    ca_chain: Optional[list[str]] = None
+    cert: Optional[str] = None
+
+
+class CertificatesIntegration:
+    def __init__(self, charm: CharmBase) -> None:
+        self._charm = charm
+        self._container = charm._container
+
+        hostname = charm.config.get("hostname")
```

Comment:
I think the certificates should be recreated in case the hostname config is changed

Reply:
here we would need the hostname to be the k8s service though, right?what are the chances that would change?

not sure we should cater for each unit hostname as a `v0`, unless that is really trivial

Reply:
@natalian98 yeah, it should. However, we cannot immediately update the certificates in the `config-changed` event since the secret has not been removed yet. I was expecting the `cert-changed` event will be triggered since there will be a set of certificates. However, the CertHandler v1 does not trigger the `cert-changed` event again. I will try v0 instead.

@shipperizer yeah, it can be the internal DNS record of the k8s service. I left a configuration window for the server hostname if the end users might want some meaningful name. We can chat about this tomorrow.